Web Frontend

nateyazdani opened the pull request:
First off, let me apologize for the sheer size of this pull request. This has been brewing for quite a long time, and I had originally expected it to be a lot more lightweight. I promise that the majority of the code is either super simple or from a reputable third-party (_e.g._, jQuery, Bootstrap). Aside from that, this is pretty much ~~ripping off~~ inspired by datagrepper's website. To give you an idea of what this looks like, the home page is pictured below. (Note: I found a bug halfway through writing this, but I haven't been able to send a PR due to having such a, ahem, polluted working tree. That PR is now pending.)

![image](https://cloud.githubusercontent.com/assets/4121048/10036146/3a0578c0-6158-11e5-8a13-3094dfde53ba.png)

Commit:
Add style and font resources for web interface

CSS and font files taken from datagrepper for stylistic consistency.

Commit:
Configure Apache to serve static files

Modifies the production Apache HTTP Server configuration to serve static files
from '/usr/share/statscache/static/' as '/static/', as is compatible with Flask.

Commit:
Install resource and config. files with package

Provide options to setuptools.setup so that it will automatically install
configuration and resource files into the appropriate system directories during
package installation.

Commit:
Very basic web frontend

Provide extremely simple HTML views of model feeds and error pages.

Commit:
Add HTML, CSS, and JS files to manifest

Includes web resources with packages distribution.

Commit:
Correct frontend timestamp parsing

Convert the second-granularity timestamps into milliseconds before passing them
to the JavaScript Date constructor, which expects milliseconds.

Commit:
Correct variable name

When rebasing the experimental front-end branch onto the changes introduced by
pagination and default serialization, a variable name change conflicted with
the current code.

Commit:
Revise template hierarchy

Move more common layout structure into the base template, layout.html.

Commit:
Stylistic changes to model feed web page

Use the English plugin names in the feed web page's title and byline and use a
monospaced font to display the data rows, in order to subtly imply the "rawness"
of the information being displayed.

Commit:
Remove underscore.js dependency

It turned out not to be worth having to keep another JavaScript file around.

Commit:
Perform proper HTML escaping in model feed

Best not to have glaring security holes lying around.

Commit:
Remove redundant sorting in model feed

Before the pagination code was implemented in the web app, results from querying
a model were sorted in the JavaScript of the feed web page. As this is no longer
necessary (not to mention extremely inefficient), this code has been removed.

Commit:
Slight revision of feed JavaScript

Replace uses of Array.prototype.map() with Array.prototype.forEach() where more
appropriate and rename a variable.

Commit:
Remove unnecessary argument to error template

Commit:
Create placeholder reference web page

Commit:
Reduce font size of model feed rows

A nine-point font flows much better (as in, the table does not extend beyond any
of the header content) and allows for slightly more compact display.

Commit:
Include plugin description on model feed web page

Commit:
Reduce size of statscache name in header

Using an 'h1' tag made the name stick out too much; an 'h2' tag provides a
better balance.

Commit:
Create placeholder getting started web page

Commit:
Redirect web index requests to getting started

Redirect requests to '/' and '/web/' to the getting started web page
('/web/getting-started').

Commit:
Visual updates to model feed web page

Include plugin description, rephrase page description, and move the name of the
plugin/model into the heading of the leading paragraph instead of the byline.

Commit:
Decrease the statscache header font size (again)

Using 'h2' still made the interface feel too busy, in that it detracted
attention from the proper page heading, which comes after the header. The
statscache name in the header bar (and its byline) now uses the 'h3' tag, just
like datagrepper does.

Commit:
Create placeholder dashboard web page

Commit:
Load model feed via AJAX

In the web view for models, load the actual rows via AJAX rather than by
rendering them in the template. This will facilitate later separation of the
REST API with the web interface.

Commit:
Include Bootstrap JavaScript

Include Bootstrap's JavaScript code (version 3.3.5).

Commit:
Update Bootstrap CSS to v3.3.5

This is to ensure compatbility with the associated Bootstrap JavaScript.

Commit:
Include Bootstrap's Glyphicons font

As the name suggests, this font is actually a set of icons represented as font
glyphs. The font is included through an "@font-face" from the base template,
layout.html.

Commit:
Rename the "fonts" folder to "font"

This is more consistent with the other "static/" subdirectories, i.e., "image/",
"style/", and "script/".

Commit:
Install common Bootstrap plugins

These will be used to implement controls in the dashboard and feed view.

Commit:
Refine CSS rules in feed.css

Ensure that the custom CSS rules declared for the feed view are only applied to
the elements intended.

Commit:
Install moment.js

This is needed for a Bootstrap datetime-picker plugin.

Commit:
Install datetime-picker plugin for Bootstrap

Commit:
Reflow base template

Adds some additional spacing for logical separation between various bundles of
script/stylesheets. Also ensures that any script or stylesheet comes after the
"primary" one in its collection.

Commit:
Load Moment.js prior to the datetime-picker plugin

The datetime-picker plugin for Bootstrap requires that Moment.js be loaded
first.

Commit:
Do not provide JSON data to web feed template

As the web feed view now uses AJAX to load the JSON content of the respective
models, it is no longer necessary for that same data to be redundantly
serialized and passed to the template, as it will not even be used.

Commit:
Add loading status and error message

Include a loading status/error message label next to the spinwheel. Currently,
this label will either state merely that content is loading or that an error
occurred while doing so.

Commit:
Minor code clean-up in model feed web view

Also includes HTML-escaping the column names, just in case.

Commit:
Add missing ID to primary table element

The table for displaying the rows of the model feed was given the ID
'stats-table', but this seems to have been left out of prior commits on
accident.

Commit:
Add start/stop datetime-pickers to model feed view

The start datetime defaults to the epoch, and the stop datetime defaults to
moment at which the page was generated. Both of these times are normalized to
the next valid minute, in order to suppress the datetime picker widgets from
doing so, which triggers and unnecessary reload of the model content.

Additionally, the start datetime-picker disallows selection of datetimes before
the epoch and after the currently chosen stop datetime, with the stop datetime-
picker supporting similar behavior with respect to the current moment and the
currently chosen start datetime.

Some changes to the identification of the model table were also necessary, in
order to prevent the web page's JavaScript from interfering with the datetime-
pickers' elements.

Commit:
Suppress potential duplication of API requests

Commit:
Include jQuery "appear" plugin

This plugin triggers a jQuery event when an element comes into view, which will
be used to implement continuous scrolling of the model feed in the web view.

Commit:
Continuous scrolling in web view of model feed

Commit:
Remove unnecessary variable

Commit:
Remove unneeded Bootstrap plugins

As it turns out, the Bootstrap dropdown and button plugins do nothing and were
actually preventing dropdown menus from working at all (perhaps due to some
conflict with the latest version of Bootstrap's JavaScript).

Commit:
Add ordering dropdown menu to view of model feed

Commit:
Add comment on model feed template

Briefly explain the arithmetic on the timestamps.

Commit:
Reload model feed on ordering selection

Commit:
Suppress unnecessary model feed reloads

If the user selects an ordering that's the same as before, then don't flush the
table and reload everything.

Commit:
Remove unused CSS file

Commit:
Convert tabs to spaces (1:4)

Commit:
Initial version of 'Getting started' page

This is the initial write of the landing page for the statscache web portal.
Note that this pulls in highlight.js from the CloudFlare CDN.

Commit:
Add custom CSS rules to layout template

These rules were created for the getting started page but apply equally well to
any other page. The rules serve primarily to visually offset subsections and
asides.

Commit:
Correct conditional page width

Commit:
Correct example plugin in getting started

Commit:
Explain plugin entry-points in getting started

Commit:
Move utility JS methods to base template

Commit:
Add comment regarding widget to base template

Commit:
Initial write of API reference web page

Commit:
Fix incorrect header name in web page

Commit:
Initial version of dashboard web page

Commit:
Separate the web endpoints from the APIs

This makes the URL hierarchy more consistent, while also preventing unnecessary
processing from happening in the API route's code-path.

Commit:
Install static resources accessibly to Apache

By installing static resources in /usr/share directory, Apache can access them
to service static file requests much more efficiently than Flask.

ralphbean commented:
That frontpage looks awesome!  :p

ralphbean commented on setup.py:
```diff
@@ -20,6 +25,20 @@ def get_description():
 ]
 
+# TODO: let a flag or option determine whether to actually install these
+install_data = [
```
Yeah, watch out for this.  This makes things _really_ screwy when you try to pip install the project into a virtualenv.  We've tried doing it in a handful of other projects and always ended up backing it out and relying on .spec files to handle `/usr/share` and `/etc`.

ralphbean commented on statscache/templates/reference.html:
```diff
@@ -0,0 +1,146 @@
+            multiple convenient data formats.
+        </p>
+    </section>
+{# This page really ought to be more thorough #}
```
It's a good start.  We can revise it as we learn what people get confused about (via what questions they come and ask us).

nateyazdani replied:
Sounds good :) part of me really wanted to keep at it, but I'd already spent days on just that page (not much of a writer)

ralphbean commented:
:+1:, but I'd consider removing that `install_data` bit in setup.py.. that's my only beef.  It looks convenient, but it ends up the opposite.

nateyazdani commented:
Thanks for the feedback :) And that's a bummer about `install_data`. I was curious about how it might interact with virtualenvs, and I guess the answer is not too well :/ I'll take that out and add a note in `setup.py` to remind packagers to do soemthing like that.

Commit:
Do not install package data in setup.py

Let the specfile handle the installation of Apache configuration files and
static file resources. This prevents clobbering of system directories from
within a virtualenv.

ralphbean commented:
:+1: to merge!
